token/services/identity/msp/rolefactory.go

```diff
@@ -85,27 +85,34 @@ func NewRoleFactory(
 }
 
 // NewIdemix creates a new Idemix-based role
-func (f *RoleFactory) NewIdemix(role driver.IdentityRole, cacheSize int, issuerPublicKey []byte, curveID math3.CurveID, additionalKMPs ...common.KeyManagerProvider) (identity.Role, error) {
+func (f *RoleFactory) NewIdemix(role driver.IdentityRole, cacheSize int, issuerPublicKey *crypto.IdemixIssuerPublicKey, additionalKMPs ...common.KeyManagerProvider) (identity.Role, error) {
 	f.Logger.Debugf("create idemix role for [%s]", driver.IdentityRoleStrings[role])
-
-	backend, err := f.StorageProvider.NewKeystore()
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to get new keystore backend")
+	if issuerPublicKey == nil && len(additionalKMPs) == 0 {
+		return nil, errors.New("expected a non-nil idemix public key")
 	}
-	keyStore, err := msp.NewKeyStore(curveID, backend)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to instantiate bccsp key store")
+
+	kmps := additionalKMPs
+	if issuerPublicKey != nil {
+		backend, err := f.StorageProvider.NewKeystore()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get new keystore backend")
+		}
+		keyStore, err := msp.NewKeyStore(issuerPublicKey.Curve, backend)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to instantiate bccsp key store")
+		}
+		kmp := idemix2.NewKeyManagerProvider(
+			issuerPublicKey.PublicKey,
+			issuerPublicKey.Curve,
+			RoleToMSPID[role],
+			keyStore,
+			f.SignerService,
+			f.Config,
+			cacheSize,
+			f.ignoreRemote,
+		)
+		kmps = append([]common.KeyManagerProvider{kmp}, additionalKMPs...)
```

Comment:
The reason you didn't do `kmps = append(kmps, kmp)` is because you would modify the additionalKMPs slice reference?

Reply:
no, it is the order. The order gives priority to the different kmps